Bugfix: Make datasource HTTP requests verify TLS by default

mattbostock opened the pull request:
Make proxied datasource HTTP requests verify TLS by default and add a checkbox that allows Grafana administrators to skip verification for testing purposes. Previously, TLS verification for datasources was skipped by default.

Also, allow TLS client authentication and user-supplied TLS CA to be used independently (previously, TLS client authentication had to be enabled for either to work).

Fixes #9371, #9366, #5334, and #8812.

## Before

<img width="1000" alt="screen shot 2017-09-28 at 15 10 47" src="https://user-images.githubusercontent.com/4348848/30971229-3cd5fc4a-a45f-11e7-986a-73dbaff51490.png">


## After

TLS verification is skipped, meaning we get a 404 (which is expected for this URL):

<img width="1039" alt="screen shot 2017-09-28 at 14 55 23" src="https://user-images.githubusercontent.com/4348848/30971024-bd2c1628-a45e-11e7-90e1-88b279bf0c74.png">

TLS verification is enabled, which means we now get a 'bad gateway error':

<img width="1037" alt="screen shot 2017-09-28 at 14 55 14" src="https://user-images.githubusercontent.com/4348848/30971025-bd314896-a45e-11e7-9035-22048a5f9758.png">

I have done some brief manual testing in addition to the additional units tests but would appreciate if others using self-signed certificates and/or TLS client authentication could test this out.

I also tidied the UI a little for consistency and found an issue where an error parsing the CA PEM could go unreported (eb1c3f3).

## Todo:

- [x] Data migrations: I have not considered yet how the existing `tlsAuth` 'jsonData' key will be migrated for existing installations; help would be appreciated with that.

I didn't try to prevent an admin from setting a TLS CA when 'Skip verify' is enabled (which means that setting a CA is pointless), that should probably be tackled in a future change.

mattbostock committed:
Tests: Clarify what InsecureSkipVerify does

mattbostock committed:
Tidy spacing in datasource TLS settings

mattbostock committed:
Verify datasource TLS and split client auth and CA

mattbostock committed:
Alias macron package in app_routes.go

...to make this file compatible with goimports:
https://godoc.org/golang.org/x/tools/cmd/goimports

mattbostock committed:
Make URL capitalisation consistent in UI

URL is an acronym, it should be all caps.

mattbostock committed:
Datasource HTTP settings: Add TLS skip verify

In c04d95f35 I changed the default for datasource HTTP requests so that
TLS is always verified.

This commit adds a checkbox to allow an admin to explicitly skip TLS
verification, for testing purposes.

mattbostock committed:
Datasource settings: Make HTTP all caps

It's an acronym, so it should be all caps.

mattbostock committed:
Return error if datasource TLS CA not parsed

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/grafana?pullRequest=9377) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/grafana?pullRequest=9377) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/grafana?pullRequest=9377) it.</sub>

mattbostock commented:
/cc @bobrik @bergquist

mattbostock commented:
For other places where TLS is not verified (other than datasources), see: #9378.

mattbostock commented:
@bergquist: Will get back to you ASAP re: the CLA.

mattbostock commented:
@bergquist: In the meantime, can you please comment on my 'todo' item regarding data migrations? the `tlsAuth` key was renamed so this could perhaps break existing installations?

bergquist commented:
@mattbostock yes it would. 

We don't provide any way of migrating fields in .jsonData so in this case I would question the need for renaming it. The new name is **better**  but backwards comp ability is more important. 

The other option would be to read from the old value first and override with the new value if it exists. But "hacks" like that would have to stay for a long time.

mattbostock commented:
Thanks @bergquist.

I think I'll check for the old value first so that we can use the clearer naming going forwards to avoid any potential future confusion, if that sounds okay to you.

bergquist commented:
@mattbostock thats ok with me. `public/app/features/plugins/ds_edit_ctrl.ts` should migrate the value when people load the edit data source page.

mattbostock commented:
@bergquist: Updating the TypeScript file is beyond my JS competency within the time I have. If there's no easier way (or an example I could follow) I'll continue to use the old naming in the interests of getting this merged sooner (CLA pending of course).

mattbostock committed:
Retain old name for TLS client auth

I renamed `tlsAuth` to `tlsClientAuth` to better describe the fact that
this variable is used to enable TLS client authentication (as opposed to
server authentication) in c04d95f35.

However, changing the name breaks backwards compatibility for existing
installations using this feature and Grafana does not have a
standardised way of migrating changes in the schema:
https://github.com/grafana/grafana/pull/9377#issuecomment-333063543

For reasons of expediency given the severity of the bug (not verifying
TLS), keep the old name.

mattbostock commented:
Reverted to the old name in 7e1c9e0.

mattbostock commented:
@bergquist: Worth stating that this change will still break existing installations if they're relying on the fact that TLS verification was disabled. The fix for them is to check the 'Skip TLS verification' checkbox on every datasource affected.

I don't think there's any reasonable workaround for that and people should be encouraged to review their security settings IMHO in case they were unaware that TLS was not being verified.

bergquist commented:
@mattbostock I really dislike imposing breaking changes on our users. 

But in this case I agree with you. Still waiting for the CLA signature :)

mattbostock commented:
@bergquist: I've signed the CLA, was waiting for confirmation from my employer.

bergquist commented:
@torkelo Please review this. It looks good to me but I think it would be good with one extra pair of eye.

torkelo commented on public/app/features/plugins/partials/ds_http_settings.html:
```diff
@@ -30,35 +30,29 @@ <h3 class="page-heading">Http settings</h3>
+        <gf-form-switch class="gf-form" label="With Credentials" tooltip="Whether credentials such as cookies or auth headers should be sent with cross-site requests." checked="current.withCredentials" label-class="width-11" switch-class="max-width-6"></gf-form-switch>
+      </div>
+      <div class="gf-form-inline">
+        <gf-form-switch class="gf-form" ng-if="current.access=='proxy'" label="TLS Client Auth" label-class="width-8" checked="current.jsonData.tlsClientAuth" switch-class="max-width-6"></gf-form-switch>
```
looks like this breaking changing in changing the the jsonData property name from jsonData.tlsAuth to jsonData.tlsClientAuth or am I missing something? wont this break for everyone who has jsonData.tlsAuth enabled before?

torkelo commented:
think I found a breaking change in the renaming of jsonData.tlsAuth to jsonData.tlsClientAuth

mattbostock commented:
@torkelo: Fixed in 4f3856a.

torkelo commented:
great!

torkelo commented on pkg/models/datasource_cache.go:
```diff
@@ -45,9 +46,16 @@ func (ds *DataSource) GetHttpTransport() (*http.Transport, error) {
 
+	var tlsSkipVerify, tlsClientAuth, tlsAuthWithCACert bool
+	if ds.JsonData != nil {
+		tlsClientAuth = ds.JsonData.Get("tlsAuth").MustBool(false)
```
maybe rename this variable to tlsAuth as well to make the naming more consistent

mattbostock replied:
I really wanted to rename `tlsAuth` to `tlsClientAuth`, but there's no good migration path currently.

I kept `tlsClientAuth` in the Go code since it could be changed without breaking existing installations and makes the intent clearer in the code. I'd prefer to keep it as-is to avoid any future regressions.

bergquist commented:
Looks good! Merging this to master. This will be available in the 4.6.0-beta1 release

mattbostock commented on pkg/models/datasource_cache.go:
```diff
@@ -62,30 +70,24 @@ func (ds *DataSource) GetHttpTransport() (*http.Transport, error) {
-			if ok {
-				transport.TLSClientConfig.RootCAs = caPool
+			if !ok {
+				return nil, errors.New("Failed to parse TLS CA PEM certificate")
```
This error should start with a lowercase letter, I'll raise a new PR to fix.

jcberthon commented:
Maybe I misunderstood the option, but it does not seem to work for me.

I'm using a custom (internal, i.-e. non-public) CA. I've signed the InfluxDB certificate with this custom CA. Using `curl` and referencing the CA public key, I can successfully connect to the InfluxDB DB.

So I've ticked "With CA cert" and in the "TLS Auth Cert" field I copied first only the public root CA key. I tried, it did not work. I then copied both the intermediate and root public keys "bundled". It did not work better. I've finally unticked the "With CA Cert" and installed at OS level both CA public keys, it now works.

But I'm still puzzled why it did not work, or maybe I simply misunderstood the meaning of the "With CA cert" option.

Could someone clarify?

ctron commented:
I have the same problem as @jcberthon reported. To me it looks like there is currently no way to specify a custom CA cert to connecting to a datasource.
